Map changing events and map change state management

tallytalwar opened the pull request:
- Provides access to platform code to check if camera is in an animating state.
- Adds logic to use this camera animating state to aptly invoke map changing MapChangeListener interface methods for both iOS and android platforms.
- [Android] Adds MapChangeStateManager to aptly manage map change states (`will update`, `will update animated`, `is updating`, `did update`, `did update animated`).
- [x] Implement the above mentioned StateManager for iOS
- [x] implement state change notifier for iOS

This also fixes (3ec875c) a regression which was causing make `DID_CHANGE` to be called repeated irrespective of a cameraAnimationCallback being set or not. (https://github.com/tangrams/tangram-es/pull/1853/commits/0908a17cd8a556b57a075f77ada21dfaf784863c)

matteblair commented on platforms/android/tangram/src/main/java/com/mapzen/tangram/MapController.java:
```diff
@@ -1089,24 +1088,66 @@ public void removeAllMarkers() {
+
+        enum MapRegionChangeState {
+            MAP_REGION_NO_CHANGE,
+            MAP_REGION_WILL_CHANGE,
```
Do you think it would be easier to think of change events as transitions rather than states? The states you have here seem to me like transitions between "idle", "animating", and "jumping".

tallytalwar replied:
Hmm, it might be actually, will give it a try.

tallytalwar replied:
Gave this some more thought, I think it is simpler and more logical to use transitions instead of the `idle`, `animating` and `jumping` states.
Transitions seems to match exactly with the map events we are catering to here, also can help reduce some logic, example:
- One should not repeated call `mapChangeListener.onRegionWillChange(animated)` is the transition is already `WILL_CHANGE_ANIMTED` or `WILL_CHANGE`. I do not think this can be aptly mapped with the states.

tallytalwar replied:
I can however change the enum name to `MapRegionChangeTransitions`?

matteblair replied:
It was easy enough to implement on top of your changes so I wrote out the logic as I've been thinking of it: https://github.com/tangrams/tangram-es/commit/15e1859cb1ddbef181b8cf8b5275bbb9a9eb7d92

tallytalwar replied:
Hmm I see, will see if I can extend that logic. Example if I start panning  the map when map was animating say due to a flyTo, or if some event explicitly sets a position (basically state set to jumping) when map is animating. Such state management logic will have to be added to this.

matteblair replied:
Not claiming that this is objectively better in any way, I just wanted to see how it would work.

tallytalwar replied:
Ohh totally, I actually like lesser enums (states vs transitions) getting us to the same place. So will be a cleaner approach.

tallytalwar replied:
@matteblair adopted States here: 53fb10f3ed0e1d64a4a1d33a617556c9a5e3f000

matteblair commented on platforms/android/tangram/src/main/java/com/mapzen/tangram/MapController.java:
```diff
@@ -1089,24 +1098,56 @@ public void removeAllMarkers() {
-    private void onRegionWillChange(boolean animated) {
-        if (mapChangeListener != null) {
-            mapChangeListener.onRegionWillChange(animated);
+    private static class MapRegionChangeStateManager {
```
Why create an inner class here? Seems to create clutter and I'm not sure what's gained from it.

tallytalwar replied:
I wanted to encapsulate the logic and not make it part of MapController itself. Since its internal, maybe I can add a package private external class for the same.

matteblair replied:
There's not much to encapsulate. I think the worst that we could do is accidentally set the state directly instead of using the method - and actually we can do that anyway even with the inner class.

matteblair commented on platforms/android/tangram/src/main/java/com/mapzen/tangram/MapController.java:
```diff
@@ -1089,24 +1098,56 @@ public void removeAllMarkers() {
-    private void onRegionDidChange(boolean animated) {
-        if (mapChangeListener != null) {
-            mapChangeListener.onRegionDidChange(animated);
+        WeakReference<MapChangeListener> mapChangeListenerWeakReference;
```
Why is this a weak reference?

tallytalwar replied:
To me it didn't make sense for MapRegionChangeManager to hold a strong reference to MapChangeListener. Also during MapController init, we need to initialize MapRegionChangeManager with a `null`!

matteblair replied:
Also not an issue if we remove the inner class..

tallytalwar replied:
Inner class is gone => weak reference is gone!

matteblair commented on platforms/android/tangram/src/main/java/com/mapzen/tangram/MapController.java:
```diff
@@ -1089,24 +1098,56 @@ public void removeAllMarkers() {
+                    case JUMPING:
+                        if (state == MapRegionChangeState.IDLE) {
+                            mapChangeListener.onRegionDidChange(false);
+                        } else if (state == MapRegionChangeState.ANIMATING) { // e.g. panning to fling
```
I'm not sure that `JUMPING` -> `ANIMATING` is a valid transition. This would mean we produce "is changing" calls with no preceding "will change animated", which seems wrong. Also, panning currently sets the state to `ANIMATING` (though maybe it _should_ be `JUMPING`).

tallytalwar replied:
I feel these gesture responders should be `ANIMATING`.

tallytalwar replied:
`JUMPING` state should imply any change which results in sudden change in map view:
eg. a cameraUpdate call with duration set to 0!

matteblair commented on platforms/android/tangram/src/main/java/com/mapzen/tangram/MapController.java:
```diff
@@ -1089,24 +1098,56 @@ public void removeAllMarkers() {
+                            mapChangeListener.onRegionDidChange(true);
+                        } else if (state == MapRegionChangeState.ANIMATING) {
+                            mapChangeListener.onRegionIsChanging();
+                        } else { // ?? ANIMATING -> JUMPING?? Possible?
```
I'd prefer `else if (state == MapRegionChangeState.JUMPING)` just to be clearer, and in case we add more states.

matteblair replied:
I don't _think_ this transition is ever triggered - animation should be canceled and set the state to `IDLE` before it becomes `JUMPING`. I guess we could find out for sure by putting a breakpoint here or throwing an exception.

tallytalwar replied:
Yeah I was not able to trigger this at all. Wanted to keep it in there for discussion point of view.

tallytalwar replied:
Going to remove this. Do not think it makes sense to respect this state transition of `ANIMATING`->`JUMPING`. The transitions of `DID_ANIMATING`, `WILL_CHANGE` implying an `IDLE` state in between will always happen.

matteblair commented on platforms/android/tangram/src/main/java/com/mapzen/tangram/MapController.java:
```diff
@@ -1380,9 +1422,14 @@ public void onDrawFrame(final GL10 gl) {
+        }
+
+        if (isCameraEasing) {
+            mapRegionChangeStateManager.setMapRegionState(MapRegionChangeStateManager.MapRegionChangeState.ANIMATING);
```
This executes a listener callback on the GL thread, we should run it on the UI thread instead.

matteblair commented on platforms/android/tangram/src/main/java/com/mapzen/tangram/MapController.java:
```diff
@@ -85,6 +85,12 @@
         SELECTION_BUFFER,
     }
 
+    public enum MapRegionChangeState {
```
Think this enum should be private? No need for it to be visible outside this class.

matteblair commented on platforms/android/tangram/src/main/java/com/mapzen/tangram/MapController.java:
```diff
@@ -1380,9 +1406,19 @@ public void onDrawFrame(final GL10 gl) {
+        }
+
+        if (isCameraEasing) {
+            uiThreadHandler.post(new Runnable() {
```
Small thing: we can make one Runnable and post it repeatedly to avoid excess GC churn.

tallytalwar replied:
Yeah totally, I feel I have caused this comment at least a couple of times now! Lesson Learnt!

tallytalwar commented:
I am going to add ios work to the same PR. So do not merge.

tallytalwar commented:
Ok this is good for final review.

Updated with following since last discussion on this:
- fixed fling map region state change behavior on android
- Added map state change logic for iOS
- fixed a regression with cameraAnimationListener calling "DID_CHANGE" too aggressively
- Added state change notifier for a client to explicit set map region change states.

matteblair commented on platforms/ios/framework/src/TGMapView.mm:
```diff
@@ -38,6 +38,15 @@ inline CLLocationDirection convertRotationRadiansToBearingDegrees(float rotation
+/**
+ Map region change states
+ */
+typedef NS_ENUM(NSInteger, TGMapRegionChangeStates) {
```
Conventionally, enum names are singular - it makes a little more sense when using it as a type name.

matteblair commented on platforms/ios/framework/src/TGMapView.mm:
```diff
@@ -38,6 +38,15 @@ inline CLLocationDirection convertRotationRadiansToBearingDegrees(float rotation
+ Map region change states
+ */
+typedef NS_ENUM(NSInteger, TGMapRegionChangeStates) {
+    IDLE = 0,
```
Enum values should use our "TG" prefix because unlike Java everything is in the global namespace :\

Conventionally they would be something like `TGMapRegionIdle`.

tallytalwar replied:
Yeah I always forget this!!!!! Will update.

msmollin replied:
Unless you mark it as static, then its scoped to the file, but lets not do that :)

matteblair commented on platforms/ios/demo/src/MapViewController.m:
```diff
@@ -150,18 +165,32 @@ - (void)mapView:(TGMapView *)view recognizer:(UIGestureRecognizer *)recognizer d
+    TGCameraPosition* camera = [view cameraPosition];
+    camera.center = CLLocationCoordinate2DMake(coordinates.latitude, coordinates.longitude);
+    camera.zoom += 1;
+    [view setCameraPosition:camera withDuration:5000 easeType:TGEaseTypeCubic callback: ^(BOOL canceled){
```
This animation is 5000 seconds long, probably not what you intended 😄

tallytalwar replied:
Nopes, I purposely did 5000, its very soulful to take a slow route from sf to new delhi :P.

tallytalwar replied:
Also this is from a drop commit.

matteblair commented:
I'm not seeing map events during pan-flinging on iOS. Not sure why yet - are you observing this as well?

tallytalwar commented:
> I'm not seeing map events during pan-flinging on iOS. Not sure why yet - are you observing this as well?

Interesting I was able to see flinging show as animating, I specifically tested for that. Will check again, I might have refactored something later!

msmollin commented on platforms/ios/demo/src/MapViewController.h:
```diff
@@ -12,12 +12,16 @@
 
 - (void)mapView:(nonnull TGMapView *)mapView didLoadScene:(int)sceneID withError:(nullable NSError *)sceneError;
 - (void)mapViewDidCompleteLoading:(nonnull TGMapView *)mapView;
+- (void)mapViewRegionIsChanging:(TGMapView *)mapView;
```
Nit: Unnecessary to define these in the header as its implied by the `TGMapViewDelegate` protocol conformance. It doesn't hurt anything but it's unnecessary.

msmollin commented on platforms/ios/framework/src/TGMapView.h:
```diff
@@ -399,6 +399,26 @@ TG_EXPORT
+ Client can use these to trigger explicit `TGMapRegionChangeStates` changes
+ Notifies that client is going to begin map region animation
+ */
+- (void)notifyGestureDidBegin;
```
So, the general iOS guideline for callbacks is
`will` / `is` / `did` and you want to pass along the source, so normally the map view. For example:
`- (void)mapView:(TGMapView*)mapview willChangeRegion;`
`- (void)mapView:(TGMapView*)mapview regionChanged;`
`- (void)mapView:(TGMapView*)mapview didFinishChangingRegion;`

You might also want to think about, similar to gesture recognizers, something like:
`- (BOOL)shouldMapview:(TGMapView*)mapview allowRegionChange`

Another thought: I would say the naming here is a little confusing from an iOS perspective. Notification is generally done via delegate callbacks, NSNotificationCenter, or KVO. Function callbacks *do* happen, like in `viewDidLoad` and other lifecycle-ish methods, but generally it's preferable to go with one of the other 3.

Lastly, if you plan on keeping these, I would add comments here about subclassing requirements that developers *must* call super, or internal functionality will break if you don't.

msmollin replied:
Discussed this in person with @tallytalwar and understand the decisions here now :)

tallytalwar commented:
@matteblair flinging on both platforms should behave smoothly now.

matteblair commented on platforms/ios/framework/src/TGMapView.mm:
```diff
@@ -1022,15 +1050,15 @@ - (void)respondToPanGesture:(UIPanGestureRecognizer *)panRecognizer
     switch (panRecognizer.state) {
         case UIGestureRecognizerStateBegan:
-            [self regionWillChangeAnimated:YES];
+            [self setMapRegionChangeState:TGMapRegionAnimating];
```
This initially seemed like a good way to categorize these events, but it turns out that MapKit actually calls `regionWillChangeAnimated: false` at the start of panning and other gestures. To match this behavior we'll need to modify our state machine a little. I think we can just add a transition from "jumping" to itself, and call `regionIsChanging` on that transition. Then for gesture events like this we'll set the state to "jumping" repeatedly.

tallytalwar replied:
Yeah, works great. Good suggestion.

tallytalwar commented:
Ok Updated. All good here now?
